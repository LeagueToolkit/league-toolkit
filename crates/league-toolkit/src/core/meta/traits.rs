use std::io;

use super::property::{value::*, BinPropertyKind};
use enum_dispatch::enum_dispatch;

const HEADER_SIZE: usize = 5;
#[enum_dispatch]
pub trait PropertyValue {
    fn size(&self, include_header: bool) -> usize {
        self.size_no_header()
            + match include_header {
                true => HEADER_SIZE,
                false => 0,
            }
    }
    fn size_no_header(&self) -> usize;
}

pub trait ReadProperty: Sized {
    fn from_reader<R: io::Read + io::Seek>(
        reader: &mut R,
        legacy: bool,
    ) -> Result<Self, crate::core::meta::ParseError>;
}

use byteorder::ReadBytesExt as _;
pub trait ReaderExt: io::Read {
    /// Reads a u8 as a property kind
    fn read_property_kind(
        &mut self,
        legacy: bool,
    ) -> Result<BinPropertyKind, crate::core::meta::ParseError> {
        BinPropertyKind::unpack(self.read_u8()?, legacy)
    }
}

impl<R: io::Read + ?Sized> ReaderExt for R {}

pub trait WriteProperty: Sized {
    fn to_writer<R: io::Write + ?Sized>(
        &self,
        writer: &mut R,
        legacy: bool,
    ) -> Result<(), io::Error>;
}

use byteorder::WriteBytesExt as _;
pub trait WriterExt: io::Write {
    fn write_property_kind(&mut self, kind: BinPropertyKind) -> Result<(), io::Error> {
        self.write_u8(kind.into())
    }
}

impl<R: io::Write + ?Sized> WriterExt for R {}
