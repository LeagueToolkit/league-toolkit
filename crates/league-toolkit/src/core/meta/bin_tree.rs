use std::{collections::HashMap, io};

use super::{error::ParseError, BinTreeObject};

#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
#[derive(Debug)]
pub struct BinTree {
    pub is_override: bool,
    pub version: u32,

    pub objects: HashMap<u32, BinTreeObject>,
    /// List of other property bins we depend on.
    ///
    /// Property bins can depend on other property bins in a similar fashion to importing code libraries
    pub dependencies: Vec<String>,

    data_overrides: Vec<()>,
}

impl BinTree {
    pub fn new(
        objects: impl IntoIterator<Item = BinTreeObject>,
        dependencies: impl IntoIterator<Item = String>,
    ) -> Self {
        Self {
            version: 3,
            is_override: false,
            objects: objects
                .into_iter()
                .map(|o: BinTreeObject| (o.path_hash, o))
                .collect(),
            dependencies: dependencies.into_iter().collect(),
            data_overrides: Vec::new(),
        }
    }
}

impl BinTree {
    pub fn from_reader<R: io::Read + std::io::Seek>(reader: &mut R) -> Result<Self, ParseError> {
        use crate::util::ReaderExt as _;
        use byteorder::{ReadBytesExt as _, LE};

        const PROP: u32 = u32::from_le_bytes(*b"PROP");
        const PTCH: u32 = u32::from_le_bytes(*b"PTCH");

        let magic = reader.read_u32::<LE>()?;
        let is_override = match magic {
            PROP => false,
            PTCH => {
                let override_version = reader.read_u32::<LE>()?;
                if override_version != 1 {
                    return Err(ParseError::InvalidFileVersion(override_version));
                }

                // It might be possible to create an override property bin
                // and set the original file as a dependency.
                // This seems to be the object count of the override section
                let _maybe_override_object_count = reader.read_u32::<LE>()?;

                let magic = reader.read_u32::<LE>()?;
                if magic != PROP {
                    // TODO (alan): repr this in the error
                    log::error!(
                        "Expected PROP ({PROP:#x}) section after PTCH ({PTCH:#x}), got '{:#x}'",
                        magic
                    );
                    return Err(ParseError::InvalidFileSignature);
                }
                true
            }
            _ => return Err(ParseError::InvalidFileSignature),
        };

        let version = reader.read_u32::<LE>()?;
        if !matches!(version, 1..=3) {
            // TODO (alan): distinguish override/non-override version
            return Err(ParseError::InvalidFileVersion(version));
        }

        let dependencies = match version {
            2.. => {
                let dep_count = reader.read_u32::<LE>()?;
                let mut dependencies = Vec::with_capacity(dep_count as _);
                for _ in 0..dep_count {
                    dependencies.push(reader.read_len_prefixed_string::<LE>()?);
                }
                dependencies
            }
            _ => Vec::new(),
        };

        let obj_count = reader.read_u32::<LE>()? as usize;
        let mut obj_classes = Vec::with_capacity(obj_count);
        for _ in 0..obj_count {
            obj_classes.push(reader.read_u32::<LE>()?);
        }

        let mut objects = HashMap::with_capacity(obj_count);
        match Self::try_read_objects(reader, &obj_classes, &mut objects, false) {
            Ok(_) => {}
            Err(ParseError::InvalidPropertyTypePrimitive(kind)) => {
                log::warn!("Invalid prop type {kind}. Trying reading objects as legacy.");
                Self::try_read_objects(reader, &obj_classes, &mut objects, true)?;
            }
            e => e?,
        }

        let data_overrides = match (is_override, version) {
            (true, 3..) => {
                let count = reader.read_u32::<LE>()?;
                let mut v = Vec::with_capacity(count as _);
                for _ in 0..count {
                    v.push(()); // TODO: impl
                }
                v
            }
            _ => Vec::new(),
        };

        Ok(Self {
            version,
            is_override,
            objects,
            dependencies,
            data_overrides,
        })
    }

    fn try_read_objects<R: io::Read + std::io::Seek>(
        reader: &mut R,
        obj_classes: &[u32],
        objects: &mut HashMap<u32, BinTreeObject>,
        legacy: bool,
    ) -> Result<(), ParseError> {
        objects.clear();
        for &class_hash in obj_classes {
            let tree_obj = BinTreeObject::from_reader(reader, class_hash, legacy)?;
            objects.insert(tree_obj.path_hash, tree_obj);
        }
        Ok(())
    }
}
