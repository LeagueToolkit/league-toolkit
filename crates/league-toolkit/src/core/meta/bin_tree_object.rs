use std::{collections::HashMap, io};

use super::{BinProperty, ParseError};

#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
#[derive(Debug)]
pub struct BinTreeObject {
    pub path_hash: u32,
    pub properties: HashMap<u32, BinProperty>,
}

impl BinTreeObject {
    pub fn from_reader<R: io::Read>(
        reader: &mut R,
        class_hash: u32,
        legacy: bool,
    ) -> Result<Self, ParseError> {
        use byteorder::{ReadBytesExt as _, LE};

        let _size = reader.read_u32::<LE>()?;
        let path_hash = reader.read_u32::<LE>()?;

        let prop_count = reader.read_u16::<LE>()? as usize;
        let mut properties = HashMap::with_capacity(prop_count);
        for _ in 0..prop_count {
            let prop = BinProperty::from_reader(reader, legacy)?;
            log::debug!("prop: {prop:?}");
            properties.insert(prop.name_hash, prop);
        }

        Ok(Self {
            path_hash,
            properties,
        })
    }
}
