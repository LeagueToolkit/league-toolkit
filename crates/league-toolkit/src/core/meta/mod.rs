pub mod property;
pub use property::BinProperty;

mod bin_tree;
pub use bin_tree::*;

mod bin_tree_object;
pub use bin_tree_object::*;

pub mod error;
pub use error::*;

pub mod traits;
