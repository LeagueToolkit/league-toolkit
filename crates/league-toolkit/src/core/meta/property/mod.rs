use num_enum::{IntoPrimitive, TryFromPrimitive};
use std::io;
use value::PropertyValueEnum;

use super::ParseError;

pub mod value;

#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
#[derive(
    Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, IntoPrimitive, TryFromPrimitive,
)]
#[repr(u8)]
pub enum BinPropertyKind {
    // PRIMITIVE TYPES
    None = 0,
    Bool = 1,
    I8 = 2,
    U8 = 3,
    I16 = 4,
    U16 = 5,
    I32 = 6,
    U32 = 7,
    I64 = 8,
    U64 = 9,
    F32 = 10,
    Vector2 = 11,
    Vector3 = 12,
    Vector4 = 13,
    Matrix44 = 14,
    Color = 15,
    String = 16,
    Hash = 17,
    WadChunkLink = 18, // newly added

    // COMPLEX TYPES
    Container = 128,
    UnorderedContainer = 128 | 1,
    Struct = 128 | 2,
    Embedded = 128 | 3,
    ObjectLink = 128 | 4,
    Optional = 128 | 5,
    Map = 128 | 6,
    BitBool = 128 | 7,
}

impl BinPropertyKind {
    /// The WadChunkLink bin property type was newly added by Riot. For some reason they decided to put it in the middle of the enum,
    /// so we need to handle cases from before and after it existed.
    ///
    /// "Legacy" property types need to be fudged around to pretend like WadChunkLink always existed, from our pov.
    ///
    /// "Non-legacy" property types can just be used as is.
    ///
    pub fn unpack(raw: u8, legacy: bool) -> Result<BinPropertyKind, ParseError> {
        use BinPropertyKind as BPK;
        if !legacy {
            // TODO (alan): don't panic here
            return Ok(BPK::try_from_primitive(raw)?);
        }
        let mut fudged = raw;

        // if the prop type comes after where WadChunkLink is now, we need to
        if fudged >= BPK::WadChunkLink.into() && fudged < BPK::Container.into() {
            fudged -= Into::<u8>::into(BPK::WadChunkLink);
            fudged |= Into::<u8>::into(BPK::Container);
        }

        if fudged >= BPK::UnorderedContainer.into() {
            fudged += 1;
        }

        Ok(BinPropertyKind::try_from_primitive(fudged)?)
    }

    pub fn read<R: io::Read>(
        self,
        reader: &mut R,
        legacy: bool,
    ) -> Result<PropertyValueEnum, super::ParseError> {
        PropertyValueEnum::from_reader(reader, self, legacy)
    }
}

#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
#[derive(Clone, PartialEq, Debug)]
pub struct BinProperty {
    pub name_hash: u32,
    #[cfg_attr(feature = "serde", serde(flatten))]
    pub value: PropertyValueEnum,
}

impl BinProperty {
    pub fn from_reader<R: io::Read>(reader: &mut R, legacy: bool) -> Result<Self, ParseError> {
        use super::traits::ReaderExt;
        use byteorder::{ReadBytesExt as _, LE};
        let name_hash = reader.read_u32::<LE>()?;
        let kind = reader.read_property_kind(legacy)?;

        Ok(Self {
            name_hash,
            value: PropertyValueEnum::from_reader(reader, kind, legacy)?,
        })
    }
    pub fn size(&self) -> usize {
        todo!()
    }
}
