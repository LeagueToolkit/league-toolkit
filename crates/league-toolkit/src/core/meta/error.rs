use miette::Diagnostic;

use super::property::BinPropertyKind;

#[derive(Debug, thiserror::Error, Diagnostic)]
pub enum ParseError {
    #[error("Invalid file signature")]
    InvalidFileSignature,
    #[error("Invalid file version '{0}'")]
    InvalidFileVersion(u32),
    #[error("Invalid '{0}' - got '{1}'")]
    InvalidField(&'static str, String),
    #[error("Invalid property kind - {0}")]
    InvalidPropertyTypePrimitive(#[from] num_enum::TryFromPrimitiveError<BinPropertyKind>),
    #[error("Invalid size - expected {0}, got {1} bytes")]
    InvalidSize(u64, u64),

    #[error("Container type {0:?} cannot be nested!")]
    InvalidNesting(BinPropertyKind),
    #[error("Invalid map key type {0:?}, only primitive types can be used as keys.")]
    InvalidKeyType(BinPropertyKind),

    #[error(transparent)]
    ReaderError(#[from] crate::util::ReaderError),
    #[error("IO Error - {0}")]
    IOError(#[from] std::io::Error),
    #[error("UTF-8 Error - {0}")]
    Utf8Error(#[from] std::str::Utf8Error),
}
