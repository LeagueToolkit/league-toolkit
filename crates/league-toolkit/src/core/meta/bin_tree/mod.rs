use std::collections::HashMap;

mod object;
use super::error::ParseError;
pub use object::*;

pub mod read;
pub mod write;

#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
#[derive(Debug, PartialEq)]
pub struct BinTree {
    pub is_override: bool,
    pub version: u32,

    pub objects: HashMap<u32, BinTreeObject>,
    /// List of other property bins we depend on.
    ///
    /// Property bins can depend on other property bins in a similar fashion to importing code libraries
    pub dependencies: Vec<String>,

    data_overrides: Vec<()>,
}

impl BinTree {
    pub fn new(
        objects: impl IntoIterator<Item = BinTreeObject>,
        dependencies: impl IntoIterator<Item = String>,
    ) -> Self {
        Self {
            version: 3,
            is_override: false,
            objects: objects
                .into_iter()
                .map(|o: BinTreeObject| (o.path_hash, o))
                .collect(),
            dependencies: dependencies.into_iter().collect(),
            data_overrides: Vec::new(),
        }
    }
}
