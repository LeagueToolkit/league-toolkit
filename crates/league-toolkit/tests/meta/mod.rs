use std::io::Cursor;

use insta::assert_ron_snapshot;
use league_toolkit::core::meta::BinTree;
#[test]
pub fn read() {
    let mut r = Cursor::new(include_bytes!("bins/leona_small.bin"));
    let bin = BinTree::from_reader(&mut r).unwrap();
    insta::with_settings!({sort_maps => true}, {
            assert_ron_snapshot!(bin);
    });
}
