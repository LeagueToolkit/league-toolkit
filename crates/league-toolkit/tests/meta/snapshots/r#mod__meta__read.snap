---
source: crates/league-toolkit/tests/meta/mod.rs
expression: bin
---
BinTree(
  is_override: false,
  version: 3,
  objects: {
    2154231397: BinTreeObject(
      path_hash: 2154231397,
      class_hash: 1171098015,
      properties: {
        2257500010: {
          "kind": Container,
          "name_hash": 2257500010,
          "value": ContainerValue(
            items: [
              PropertyValueEnum(
                kind: Struct,
                value: StructValue(
                  class_hash: 164488258,
                  properties: {
                    431879391: {
                      "kind": U8,
                      "name_hash": 431879391,
                      "value": U8Value(0),
                    },
                    607245356: {
                      "kind": Optional,
                      "name_hash": 607245356,
                      "value": OptionalValue(F32, Some(PropertyValueEnum(
                        kind: F32,
                        value: F32Value(10.0),
                      ))),
                    },
                    895082843: {
                      "kind": BitBool,
                      "name_hash": 895082843,
                      "value": BitBoolValue(true),
                    },
                    1013213428: {
                      "kind": String,
                      "name_hash": 1013213428,
                      "value": StringValue("ASSETS/Characters/Leona/Skins/Base/Particles/common_Black.dds"),
                    },
                    1025882318: {
                      "kind": String,
                      "name_hash": 1025882318,
                      "value": StringValue("empty"),
                    },
                    1376955374: {
                      "kind": Optional,
                      "name_hash": 1376955374,
                      "value": OptionalValue(F32, Some(PropertyValueEnum(
                        kind: F32,
                        value: F32Value(1.0),
                      ))),
                    },
                    2927860839: {
                      "kind": Embedded,
                      "name_hash": 2927860839,
                      "value": EmbeddedValue(StructValue(
                        class_hash: 70254680,
                        properties: {
                          3031705514: {
                            "kind": F32,
                            "name_hash": 3031705514,
                            "value": F32Value(1.0),
                          },
                        },
                      )),
                    },
                    3043919889: {
                      "kind": String,
                      "name_hash": 3043919889,
                      "value": StringValue("ASSETS/Characters/Leona/Skins/Base/Particles/common_Black.dds"),
                    },
                  },
                ),
              ),
            ],
          ),
        },
        2624027180: {
          "kind": U16,
          "name_hash": 2624027180,
          "value": U16Value(132),
        },
        3770906030: {
          "kind": String,
          "name_hash": 3770906030,
          "value": StringValue("Play_sfx_Leona_LeonaZenithBladeMissile_hit"),
        },
        3882058040: {
          "kind": String,
          "name_hash": 3882058040,
          "value": StringValue("Characters/Leona/Skins/Skin0/Particles/Leona_Base_E_ZB_sfx_01"),
        },
        3975268028: {
          "kind": String,
          "name_hash": 3975268028,
          "value": StringValue("Leona_Base_E_ZB_sfx_01"),
        },
      },
    ),
  },
  dependencies: [],
  data_overrides: [],
)
